src/toxcore/dht/dht_friend.rs

```diff
@@ -0,0 +1,377 @@
+                                  ping_interval: u64, nodes_req_interval: u64, bad_node_timeout: u64) -> IoFuture<()> {
+        let ping_bootstrap_nodes = self.ping_bootstrap_nodes(server);
+        let ping_and_get_close_nodes = self.ping_and_get_close_nodes(server, ping_interval);
+        let send_nodes_req_random = self.send_nodes_req_random(server, bad_node_timeout, nodes_req_interval);
+
+        let res = ping_bootstrap_nodes.join3(
+            ping_and_get_close_nodes, send_nodes_req_random
+            ).map(|((),(),())| () );
+
+        self.bootstrap_nodes.nodes.clear();
+
+        Box::new(res)
+    }
+
+    fn ping_bootstrap_nodes(&mut self, server: &Server) -> IoFuture<()> {
+        let mut peers_cache = server.get_peers_cache().write();
+
+        let nodes_sender = self.bootstrap_nodes.nodes.iter()
+            .map(|node| {
+                let client = peers_cache.entry(node.pk).or_insert_with(ClientData::new);
+                server.send_nodes_req(*node, self.pk, client)
+            });
+
+        let nodes_stream = stream::futures_unordered(nodes_sender).then(|_| Ok(()));
+
+        Box::new(nodes_stream.for_each(|()| Ok(())))
+    }
+
+    // ping to close nodes of friend
+    fn ping_and_get_close_nodes(&mut self, server: &Server, ping_interval: u64) -> IoFuture<()> {
+        let mut peers_cache = server.get_peers_cache().write();
+
+        let nodes_sender = self.close_nodes.nodes.iter()
+            .map(|node| {
+                let client = peers_cache.entry(node.pk).or_insert_with(ClientData::new);
+
+                if client.last_ping_req_time.elapsed() >= Duration::from_secs(ping_interval) {
+                    client.last_ping_req_time = Instant::now();
+                    server.send_nodes_req(*node, self.pk, client)
+                } else {
+                    Box::new(future::ok(()))
+                }
+            });
+
+        let nodes_stream = stream::futures_unordered(nodes_sender).then(|_| Ok(()));
+
+        Box::new(nodes_stream.for_each(|()| Ok(())))
+    }
+
+    fn send_nodes_req_random(&mut self, server: &Server, bad_node_timeout: u64, nodes_req_interval: u64) -> IoFuture<()> {
+        let mut good_nodes = Vec::new();
+
+        let mut peers_cache = server.get_peers_cache().write();
+
+        self.close_nodes.nodes.iter()
+            .map(|node| {
+                let client = peers_cache.entry(node.pk).or_insert_with(ClientData::new);
+
+                if client.last_resp_time.elapsed() < Duration::from_secs(bad_node_timeout) {
+                    good_nodes.push(node.clone());
+                }
+            }).collect::<Vec<_>>();
+
+        let res = if !good_nodes.is_empty()
+            && self.last_nodes_req_time.elapsed() >= Duration::from_secs(nodes_req_interval)
+            && self.bootstrap_times < MAX_BOOTSTRAP_TIMES {
+
+            let num_nodes = good_nodes.len();
+            let mut random_node = random_u32() as usize % num_nodes;
+            if 0 != random_node {
```

Comment:
Add comment please that this is needed for increase probability of sending request to a close node, because this is not trivial.

Reply:
done